print-geonet.c

```diff
@@ -26,232 +27,481 @@
+#define HT_BEACON 1
+#define HT_TSB 5
+#define HT_TSB_SHB 0
+#define HT_TSB_MULTI_HOP 1
+#define ELAPSED_SECONDS 5
+
+#define IMPLEMENTED_GN_HEADER_TYPES_NUM 2
+static const u_int implemented_gn_header_types[IMPLEMENTED_GN_HEADER_TYPES_NUM] = {
+	HT_BEACON,
+	HT_TSB};
+
+static const struct tok common_header_next_header_values[] = {
+	{0, "Any"},
+	{1, "BTP-A"},
+	{2, "BTP-B"},
+	{3, "IPv6"},
+};
+
+#define HT_HST(ht, hst) (((ht) << 8) | (hst))
+static const struct tok header_type_tok[] = {
+	{HT_HST(0, 0), "Any"},
+
+	{HT_HST(1, 0), "Beacon"},
+
+	{HT_HST(2, 0), "GeoUnicast"},
+
+	{HT_HST(3, 0), "GeoAnycastCircle"},
+	{HT_HST(3, 1), "GeoAnycastRect"},
+	{HT_HST(3, 2), "GeoAnycastElipse"},
+
+	{HT_HST(4, 0), "GeoBroadcastCircle"},
+	{HT_HST(4, 1), "GeoBroadcastRect"},
+	{HT_HST(4, 2), "GeoBroadcastElipse"},
+
+	{HT_HST(5, 0), "TopoScopeBcast-SH"},
+	{HT_HST(5, 1), "TopoScopeBcast-MH"},
+
+	{HT_HST(6, 0), "LocService-Request"},
+	{HT_HST(6, 1), "LocService-Reply"},
+};
+
+static const struct tok flags_text_from_bytes[] = {
+	{0, "Stationary"},
+	{1, "Mobile"},
+};
+
+static const struct tok st_text_from_bytes[] = {
+	{1, "Pedestrian"},
+	{2, "Cyclist"},
+	{3, "Moped"},
+	{4, "Motorcycle"},
+	{5, "Passenger Car"},
+	{6, "Bus"},
+	{7, "Light Truck"},
+	{8, "Heavy Truck"},
+	{9, "Trailer"},
+	{10, "Special Vehicle"},
+	{11, "Tram"},
+	{12, "Road Side Unit"},
+};
+
+/* BasicTransportProtocol Definitions*/
+#define BTP_A 1
+#define BTP_B 2
+static const struct tok btp_port_values[] = {
+	{2001, "CAM"},
+	{2002, "DENM"},
+	{2003, "MAPEM"},
+	{2004, "SPATEM"},
+	{2005, "SAEM"},
+	{2006, "IVIM"},
+	{2007, "SREM"},
+	{2008, "SSEM"},
+	{2009, "CPM"},
+	{2010, "EVCSN_POI"},
+	{2011, "TPG"},
+	{2012, "EV_RSR"},
+	{2013, "RTCMEM"},
+	{2014, "CTLM"},
+	{2015, "CRLM"},
+	{2016, "EC_AT_REQ"},
+	{2017, "MCDM"},
+	{2018, "VAM"},
+	{2019, "IMZM"},
+	{2020, "DSM"},
+	{2021, "P2P_CRLM"},
+	{2022, "P2P_CTLM"},
+	{2023, "MRS"},
+	{2024, "P2P_FULL_CTLM"},
+	{0, NULL}};
+
 
-	/* Assuming ItsPduHeader */
-	ND_PRINT("; ItsPduHeader v:%u", GET_U_1(bp));
 
-	msg_type = GET_U_1(bp + 1);
-	ND_PRINT(" t:%u-%s", msg_type,
-	         tok2str(msg_type_values, "unknown (%u)", msg_type));
+static int is_value_in_list(u_int value, const u_int*list, u_int length_of_list){
+	for (u_int i = 0; i < length_of_list; i++){
+		if (value == list[i]){
+			return 1;
+		}
+	}
+	return 0;
 }
 
-/* EN 302 636-5-1 V2.2.1 Section 7.2: BTP-A header */
-static void
-print_btp(netdissect_options *ndo,
-	  const u_char *bp)
+
+static u_int convert_lt_to_seconds(u_int lt_base, u_int lt_multiplier)
 {
-	ND_PRINT("; BTP Dst:%u", GET_BE_U_2(bp + 0));
-	ND_PRINT(" Src:%u", GET_BE_U_2(bp + 2));
+	float base_seconds;
+	switch (lt_base)
+	{
+	case 0: // 50 milliseconds
+		base_seconds = 0.05;
+		break;
+	case 1: // 1 second
+		base_seconds = 1.0;
+		break;
+	case 2: // 10 seconds
+		base_seconds = 10.0;
+		break;
+	case 3: // 100 seconds
+		base_seconds = 100.0;
+		break;
+	default: // default to 0 second
```

Comment:
This default case is dead code (LT Base is a 2-bit value). Perhaps this should have been a comment.